lib/chai/utils/overwriteMethod.js

```diff
@@ -52,12 +52,24 @@ module.exports = function (ctx, name, method) {
   if (_method && 'function' === typeof _method)
     _super = _method;
 
-  var fn = function () {
-    var keep_ssfi = flag(this, 'keep_ssfi');
-    var old_ssfi = flag(this, 'ssfi');
-    if (!keep_ssfi && old_ssfi)
-      flag(this, 'ssfi', fn);
+  var overwritingMethodWrapper = function () {
+    // If proxy protection is disabled and this overwriting assertion hasn't
+    // been overwritten again by yet another assertion, then use this method
+    // wrapper as the starting point for removing implementation frames from the
+    // stack trace of a failed assertion.
+    //
+    // Note: If this assertion has been overwritten, and thus the `keep_ssfi`
+    // flag is set, then the overwriting method wrapper is used as the starting
+    // point instead. This prevents the overwriting method wrapper from showing
+    // up in the stack trace since it's invoked before this method wrapper.
+    if (!flag(this, 'keep_ssfi')) {
+      flag(this, 'ssfi', overwritingMethodWrapper);
+    }
 
+    // The `keep_ssfi` flag is set so that if this assertion ends up calling
+    // the overwritten assertion, then the overwritten assertion doesn't attempt
+    // to use itself as the starting point for removing implementation frames
+    // from the stack trace of a failed assertion.
     flag(this, 'keep_ssfi', true);
```

Comment:
Right, so the function stored in the `ssfi` flag serves as the second parameter for `Error.captureStackTrace`, meaning that the function itself, and all functions called after it (which are just the internals of Chai and plugins), will be removed from the stack trace if an error is thrown. In the case of property assertions, this function is actually the proxy getter, which is why this PR was needed. But in the case of method assertions, the proxy getter function completely returns before the method assertion's function is invoked, so the method assertion's function needs to be stored in the `ssfi` flag instead of the proxy getter.

But what about all functions that came before the function stored in the `ssfi` flag? Usually, the first function that comes before it is the function that the user passed as the second argument to a Mocha `it` invocation. This function contains the user's assertion that failed, so it's good that it's included in the stack trace. But there's a bunch of other functions still in the stack before this one: Mocha's internals. It turns out that these don't appear in the stack traces either because Mocha does its own manual filtering of the stack trace to get rid of their internals. It's important to remember this when troubleshooting an issue with Mocha.

(Note that because the `assert` interface acts as a wrapper around Chai assertions, there's currently an extra function call between the user's assertion and the one stored in the `ssfi` flag, so it shows in the stack trace. That's what #878 is about. It's fixable but will take some work on the `assert` interface.)

As for your final question, yes, the `keep_ssfi` flag is needed so that the overwriting function remains stored in the `ssfi` flag even after it proceeds to call the original function that it overwrote. The first function after the user's assertion is the one that needs to be in the `ssfi` flag in order for this to all work correctly.